### PR TITLE
Add an app-wide context for global props.

### PR DESCRIPTION
**Pull request by grampelberg**
We've been passing the `api` object down from the top of the react tree. With
16.x, there's now the ability to have context that can inject anywhere in the
tree. This creates a top level context provider that contains most of the global
variables we've been using (api, appData, ...). It subsequently cleans up some
of the routes and nested components.

- Bumps `react-dom` to 16.3.2 (to match `react`).
- Adds `enzyme-context-patch` for now. This is fixed in enzyme master, but there
  has not been a release yet. Needs to be removed when that is fixed.

### CODE REVIEW

**Comment by rmars** [comment]
This is pretty cool! Looking forward to not having to pass down props everywhere.

Hmm, this branch doesn't appear to be routing correctly. For example, when I navigate to the emojivoto namespaces page (by clicking the sidebar or by URL) http://localhost:8084/namespaces/emojivoto, I see a message for the default namespace.

screen shot 2018-05-14 at 1 09 12 pm

File: web/app/js/index.js [hunk 1]
```diff
@@ -21,37 +22,45 @@ if (proxyPathMatch) {
 
 let controllerNs = appData.controllerNamespace || "conduit";
 
-let api = ApiHelpers(pathPrefix);
+const context = {
+  api: ApiHelpers(pathPrefix),
+  controllerNamespace: controllerNs,
```
Comment (rmars) [inline web/app/js/index.js]: controllerNamespace does not appear to be propagated:
when I deploy the `emojivoto` demo app and look at the service mesh page for example, I see web pods showing up in the control plane table that aren't from the conduit namespace, indicating the controllerNamespace prop isn't getting set correctly.

Comment (grampelberg) [reply web/app/js/index.js]: Ahhhhha, that explains it. Thank you. Lemme fix that up then =)

File: web/app/js/index.js [hunk 2]
```diff
@@ -19,39 +21,46 @@ if (proxyPathMatch) {
   pathPrefix = proxyPathMatch[0];
 }
 
-let controllerNs = appData.controllerNamespace || "conduit";
+if (_.isEmpty(appData.controllerNamespace)) appData.controllerNamespace = 'conduit';
```
[the remark this answers is not in the document]
Comment (grampelberg) [reply web/app/js/index.js]: Curly braces on the `if` statement?

**Comment by grampelberg** [comment]
@rmars IMHO the biggest win here is around the `withREST` HOC. Because we're using context, you don't need to explicitly pass `api` into a component and wonder if it needs it or not.

**Pull request merged**